### PR TITLE
Authenticated package source support

### PR DESCRIPTION
**Pull request by infin8x**
This PR adds a few new options to the `setup-dotnet` task that can be used to authenticate to package sources. It works with both GPR and other private registries that support username/password authentication.

### CODE REVIEW

File: src/authutil.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  console.log(
+    `dotnet-auth: Finding any source references in ${existingFileLocation}, writing a new temporary configuration file with credentials to ${tempFileLocation}`
+  );
+  let xml: xmlbuilder.XMLElement;
+  let gprKeys: string[] = [];
+  let owner: string = core.getInput('owner');
+  if (!owner && feedUrl.indexOf('nuget.pkg.github.com') > -1) {
+    owner = github.context.repo.owner;
+  }
+  let sourceUrl: string = 'https://nuget.pkg.github.com/' + owner;
```
Comment (xt0rted) [inline src/authutil.ts]: This should be `https://nuget.pkg.github.com/{owner}/index.json`

Comment (infin8x) [reply src/authutil.ts]: 🤦‍♂ thanks. Somehow that got lost.

**Comment by infin8x** [comment]
Example working Actions run: https://github.com/infin8x/starter-packages/runs/247188839

**Comment by infin8x** [comment]
Key files to review:
- __tests__/authutil.test.ts (tests)
- __tests__/authutil.test.ts.snap (test case outputs)
- src/authutil.ts (all the authentication logic)
- src/setup-dotnet.ts (main file invoked by Actions, triggers authutil)

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
+{
```
Comment (markphelps) [inline .vscode/launch.json]: may want to add `.vscode` to .gitignore

File: action.yml [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
 inputs: 
   dotnet-version:
     description: 'SDK version to use. E.g. 2.2.104'
-# Deprecated option, do not use. Will not be supported after October 1, 2019
+  source-url:
+    description: 'Optional package source for which to set up authentication. Will consult any existing NuGet.config in the root of the repo and provide a temporary NuGet.config using the NUGET_AUTH_TOKEN environment variable as a ClearTextPassword'
+  owner:
+    description: 'Optional OWNER for using packages from GitHub Package Registry organizations/users other than the current repository''s owner. Only used if a GPR URL is also provided in source-url'
+  config-file:
+    description: 'Optional NuGet.config location, if your NuGet.config isn''t located in the root of the repo.'
```
Comment (markphelps) [inline action.yml]:
```suggestion
    description: 'Optional NuGet.config location, if your NuGet.config isn't located in the root of the repo.'
```

File: action.yml [hunk 4]
```diff
@@ -1,10 +1,16 @@
...
+  # Deprecated option, do not use. Will not be supported after October 1, 2019
```
Comment (markphelps) [inline action.yml]: should we just remove this then since its after 10/1?

File: action.yml (hunk 3 above)
Comment (infin8x) [reply action.yml]: This is intentional - it's how YAML does escaping.

**Comment by xt0rted** [comment]
I've been trying to get gpr package restore working on actions for over a month and a half with no success. I just tested out this branch of the action to see if it helped at all and nuget is still failing to restore my private packages. I'm not sure if it's a config issue, actions issue, or a gpr issue. This is the workflow I'm using:

```yaml
steps:
  - name: Checkout Repo
    uses: actions/checkout@master

  - name: Setup .NET Core
    uses: infin8x/setup-dotnet@master
    with:
      dotnet-version: 2.2.401
      source-url: https://nuget.pkg.github.com
    env:
      NUGET_AUTH_TOKEN: ${{secrets.GITHUB_TOKEN}}

  - name: Setup Nuget.exe
    uses: warrenbuckley/Setup-Nuget@v1

  - name: Version Info
    run: |
      nuget help
      nuget sources
      dotnet --version
    shell: cmd

  - name: NuGet Restore
    run: nuget restore
    shell: cmd
```

The output from `nuget sources` shows that the gpr endpoint was setup:

```
NuGet Version: 5.3.0.6260
  1.  Source [Enabled]
      https://nuget.pkg.github.com/paramax/index.json
  2.  nuget.org [Enabled]
      https://api.nuget.org/v3/index.json
  3.  Microsoft Visual Studio Offline Packages [Enabled]
      C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
```

But when running `nuget restore` the packages on gpr continue to fail with the following:

```
Restoring NuGet package Synergy.Core.3.0.11018.1830.
  GET https://nuget.pkg.github.com/paramax/download/synergy.core/3.0.11018.1830/synergy.core.3.0.11018.1830.nupkg
  GET https://api.nuget.org/v3-flatcontainer/synergy.core/3.0.11018.1830/synergy.core.3.0.11018.1830.nupkg
  NotFound https://api.nuget.org/v3-flatcontainer/synergy.core/3.0.11018.1830/synergy.core.3.0.11018.1830.nupkg 111ms
WARNING: NAME_UNKNOWN:nuget package "synergy.core" does not exist under owner "paramax"
  NotFound https://nuget.pkg.github.com/paramax/download/synergy.core/3.0.11018.1830/synergy.core.3.0.11018.1830.nupkg 172ms
WARNING: Unable to find version '3.0.11018.1830' of package 'Synergy.Core'.
  C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\: Package 'Synergy.Core.3.0.11018.1830' is not found on source 'C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\'.
  https://api.nuget.org/v3/index.json: Package 'Synergy.Core.3.0.11018.1830' is not found on source 'https://api.nuget.org/v3/index.json'.
  https://nuget.pkg.github.com/paramax/index.json: Package 'Synergy.Core.3.0.11018.1830' is not found on source 'https://nuget.pkg.github.com/paramax/index.json'.
```

Support deleted my packages a couple days ago and I just republished them (there was an issue processing them early on). I'm seeing the packages on the site and in the Visual Studio package explorer. They restore fine locally and on Azure DevOps but not on GitHub Actions.

**Comment by infin8x** [comment]
That last line worries me. Looking at it, everything seems correct, and it seems like the token is making it through correctly because you're not getting 401s. I'd be curious to have him try a `dotnet restore` instead and see if that works. If it does, it means GPR is missing some API that `NuGet.exe` is trying to use to restore packages.

/cc @whitneyimura - maybe someone from the GPR team can help troubleshoot?

**Comment by xt0rted** [comment]
@infin8x only my unit test projects can be restored with `dotnet restore` (the main application is full framework and doesn't use the sdk style projects). Doing a test with this though shows an error running `dotnet restore` as well.

```
error NU1102: Unable to find package Synergy.Core with version (>= 3.0.11018.1830)
```

If I clear my caches and run `nuget restore` locally the package restores successfully :confused:

**Comment by xt0rted** [comment]
I setup a test repo that's pulling in a [public nuget package](https://github.com/xt0rted/Nancy.Validation.FluentValidation.Net45/packages/42538) from gpr and that's working so it's just the private packages that are the issue (or maybe packages from an org vs. a user?). This doesn't look to be a `setup-dotnet` issue 👍 but maybe something with the actions environment. A sample run can be seen [here](https://github.com/xt0rted/actions-gpr-test/runs/272680881) which verifies your changes work 😄

**Comment by xt0rted** [comment]
> @xt0rted Is `Synergy.Core` published to a different repository than where the workflow is being run?

@dinahshi yes the `Synergy.*` packages are in a different repo under the same org. I'm going to assume that this is the most common setup when using private GPR feeds. Using my own PAT fixes it now, my original tests didn't work but the one just now did.

One of the main reasons I was moving to Actions from DevOps was so I could do away with the extra user account on this org that exists solely for CI build PATs. That adds complexity and wastes one of our users (I won't use my account for this because the PATs would give access to things outside this org, and if I'm removed from the org then all CI builds instantly break). It's disappointing to hear this process isn't simplified by moving to Actions. It'd be great if there was a setting to allow the `GITHUB_TOKEN` to have read access to the GPR feeds in the same org.

I guess an additional piece of feedback I have for this PR is to make sure there's a note saying `NUGET_AUTH_TOKEN` should be set to a user's PAT if you're trying to use private GPR feeds.

**Comment by AraHaan** [comment]
Nice; will it work on adding package sources that I manually add to places like appveyor for things like analyzers that are not in nuget.org yet?

like for example on my CI's I do ``nuget sources Add -Name "Generic Roslyn Analyzers" -Source https://dotnet.myget.org/F/roslyn-analyzers/api/v3/index.json``.

**Comment by StanleyGoldman** [comment]
@AraHaan Sorry for the delayed response. It definitely should, if not please open an issue.

**Pull request merged**